### PR TITLE
Do not print similar threads multiple times

### PR DESCRIPTION
**Pull request by kokosing**
Do not print similar threads multiple times

### CODE REVIEW

**Comment by kokosing** [comment]
Example output:
```
2019-08-19T00:01:17.139-0500 WARNING No test started or completed in 5.00s. Running tests:
	io.prestosql.tests.TestLocalQueries running for 49.37s
	io.prestosql.tests.TestLocalQueries::testLargeIn running for 8.14s.

Thread Dump:
There are 11 threads: [local-query-runner-executor-10, local-query-runner-executor-9, local-query-runner-executor-8, local-query-runner-executor-7, local-query-runner-executor-6, local-query-runner-executor-5, local-query-runner-executor-4, local-query-runner-executor-3, local-query-runner-executor-2, local-query-runner-executor-1, local-query-runner-executor-0] that look similar to:
"local-query-runner-executor-10" Id=27 TIMED_WAITING on java.util.concurrent.SynchronousQueue$TransferStack@392ff618
	at sun.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.SynchronousQueue$TransferStack@392ff618
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:460)
	at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
	at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1073)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	...

"Abandoned connection cleanup thread" Id=15 TIMED_WAITING on java.lang.ref.ReferenceQueue$Lock@3d1065e5
	at java.lang.Object.wait(Native Method)
	-  waiting on java.lang.ref.ReferenceQueue$Lock@3d1065e5
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	at com.mysql.jdbc.AbandonedConnectionCleanupThread.run(AbandonedConnectionCleanupThread.java:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@4b9df8a

"local-query-runner-scheduler-0" Id=14 TIMED_WAITING on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@7812fc23
	at sun.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@7812fc23
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	...

"FinalizerService" Id=13 WAITING on java.lang.ref.ReferenceQueue$Lock@ff9d912
	at java.lang.Object.wait(Native Method)
	-  waiting on java.lang.ref.ReferenceQueue$Lock@ff9d912
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at io.prestosql.util.FinalizerService.processFinalizerQueue(FinalizerService.java:92)
	at io.prestosql.util.FinalizerService$$Lambda$135/221634215.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@7770f470

"TestHangMonitor" Id=12 RUNNABLE
	at sun.management.ThreadImpl.dumpThreads0(Native Method)
	at sun.management.ThreadImpl.dumpAllThreads(ThreadImpl.java:454)
	at io.prestosql.tests.LogTestDurationListener.dumpAllThreads(LogTestDurationListener.java:125)
	at io.prestosql.tests.LogTestDurationListener.checkForTestHang(LogTestDurationListener.java:112)
	at io.prestosql.tests.LogTestDurationListener$$Lambda$1/644460953.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@5afa04c

"Monitor Ctrl-Break" Id=5 RUNNABLE (in native)
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	-  locked java.io.InputStreamReader@49a061c3
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	...

"Signal Dispatcher" Id=4 RUNNABLE

"Finalizer" Id=3 WAITING on java.lang.ref.ReferenceQueue$Lock@77a11a5f
	at java.lang.Object.wait(Native Method)
	-  waiting on java.lang.ref.ReferenceQueue$Lock@77a11a5f
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)

"Reference Handler" Id=2 WAITING on java.lang.ref.Reference$Lock@183d036c
	at java.lang.Object.wait(Native Method)
	-  waiting on java.lang.ref.Reference$Lock@183d036c
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"main" Id=1 RUNNABLE
	at java.util.stream.StreamOpFlag.fromCharacteristics(StreamOpFlag.java:733)
	at java.util.stream.StreamSupport.stream(StreamSupport.java:70)
	at java.util.Collection.stream(Collection.java:581)
	at java.util.Collections$UnmodifiableCollection.stream(Collections.java:1094)
	at io.prestosql.metadata.SignatureBinder.checkNoLiteralVariableUsageAcrossTypes(SignatureBinder.java:197)
	at io.prestosql.metadata.SignatureBinder.checkNoLiteralVariableUsageAcrossTypes(SignatureBinder.java:191)
	at io.prestosql.metadata.SignatureBinder.<init>(SignatureBinder.java:82)
	at io.prestosql.metadata.FunctionRegistry.identifyApplicableFunctions(FunctionRegistry.java:777)
	...
```

File: presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java [hunk 1]
```diff
@@ -117,10 +121,25 @@ else if (finished.get()) {
 
     private static void dumpAllThreads(String message)
     {
-        LOG.warn("%s\n\nFull Thread Dump:\n%s", message,
-                Arrays.stream(getThreadMXBean().dumpAllThreads(true, true))
-                        .map(ThreadInfo::toString)
-                        .collect(joining("\n")));
+        ListMultimap<String, ThreadInfo> threadInfos = Multimaps.index(
+                Arrays.asList(getThreadMXBean().dumpAllThreads(true, true)),
+                threadInfo -> Arrays.toString(threadInfo.getStackTrace()));
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java]: We used `ThreadInfo::toString`, now we print only `getStackTrace`. This won't include locks information.

File: presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java [hunk 2]
```diff
@@ -117,10 +121,25 @@ else if (finished.get()) {
...
+        LOG.warn("%s\n\nThread Dump:\n%s", message,
+                threadInfos.keySet().stream()
+                        .map(threadStack -> {
+                            List<ThreadInfo> similarThreads = threadInfos.get(threadStack);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java]: iterate over `entrySet()` or `asMap().entrySet()`

File: presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java [hunk 3]
```diff
@@ -117,10 +121,25 @@ else if (finished.get()) {
...
+                            checkState(!similarThreads.isEmpty());
+                            if (similarThreads.size() == 1) {
+                                return similarThreads.get(0).toString();
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java]: `getOnlyElement()`

File: presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java]: `entrySet()` unrolls the multimap

File: presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java]: Yes. This is the cost. Do you find this useful?

**Pull request closed without merging**